chain-tx-enclave-next/mls/src/tree.rs

```diff
@@ -372,8 +317,12 @@ impl TreePublicKey {
                     let left_pos = px.left();
                     let right_pos = px.right(leafs_len);
                     let parent_hash = match (
-                        nodes[left_pos.node_index()].parent_hash(),
-                        nodes[right_pos.node_index()].parent_hash(),
+                        nodes[left_pos.node_index()]
+                            .as_ref()
+                            .and_then(|node| node.parent_hash()),
+                        nodes[right_pos.node_index()]
+                            .as_ref()
+                            .and_then(|node| node.parent_hash()),
                     ) {
                         (None, Some(hash)) => hash,
                         (Some(hash), None) => hash,
```

Comment:
it was there already before -- maybe it's correct, but just re-reading that

Reply:
>For each non-empty parent node, verify that exactly one of the node's children are non-empty and have the hash of this node set as their parent_hash value (if the child is another parent) or has a parent_hash extension in the KeyPackage containing the same value (if the child is a leaf).

Reply:
looks indeed problematic, I'm reading that too.

Reply:
fixed the parent_hash compare.

Reply:
can you add a test that would previously fail with this bug, but works now?

Reply:
I don’t think so, the previous code is comparing two unrelated hashes, it takes a hash collision to make them equal I guess?

Reply:
I'm just puzzled why it "worked" before (i.e. why the unit tests didn't fail), as it had an opposite if condition (returning "ParentHashDontMatch" if they matched) -- or was it that the parent hash computation wasn't correct, so even though the condition was opposite, it always passed?

Reply:
> I'm just puzzled why it "worked" before (i.e. why the unit tests didn't fail), as it had an opposite if condition (returning "ParentHashDontMatch" if they matched) -- or was it that the parent hash computation wasn't correct, so even though the condition was opposite, it always passed?

Because it was comparing two unrelated hash (it was using `node_hash` which is not `parent_hash`), so it's always not equal, that combined with the opposite if condition, the code runs through.

Reply:
https://github.com/crypto-com/chain/issues/2066